src/snowflake/snowpark/_internal/temp_table_cleaner.py

```diff
@@ -0,0 +1,78 @@
+#
+
+import logging
+import threading
+from collections import defaultdict
+from queue import Queue
+from typing import TYPE_CHECKING, Dict
+
+if TYPE_CHECKING:
+    from snowflake.snowpark.session import Session  # pragma: no cover
+
+
+class TempTableCleaner:
+    """
+    Clean up temporary tables created automatically by Snowpark at the session level.
+
+    This function identifies and removes any temporary tables that have been
+    generated by Snowpark during the current session. Temporary tables are
+    typically used for intermediate computations (e.g., df.cache_result) and
+    are not needed when they are not referenced anymore. Removing these tables
+    helps maintain a clean working environment and reduce storage cost for a long-running
+    session.
+    """
+
+    def __init__(self, session: "Session") -> None:
+        self.session = session
+        # this dict maintains key-value pair from Snowpark-generated temp table fully-qualified name
+        # to its reference count for later temp table management
+        # this dict will still be maintained even if the cleaner is stopped (`stop()` is called)
+        self.ref_count_map: Dict[str, int] = defaultdict(int)
+        self.queue = Queue()
+        self.cleanup_thread = None
+
+    def inc_ref_count(self, name: str) -> None:
+        self.ref_count_map[name] += 1
+
+    def dec_ref_count(self, name: str) -> None:
+        self.ref_count_map[name] -= 1
+        if self.ref_count_map[name] <= 0:
+            self.ref_count_map.pop(name)
+            # clean up
+            if self.is_alive():
+                self.queue.put(name)
+
+    def process_cleanup(self) -> None:
+        while True:
+            table_name = self.queue.get()
+            if table_name is None:
+                # exit the thread loop
+                break
+            self.drop_table(table_name)
+
+    def drop_table(self, name: str) -> None:
+        common_log_text = f"temp table {name} in session {self.session.session_id}"
+        try:
+            # TODO SNOW-1556553: Remove this workaround once multi-threading of Snowpark session is supported
+            with self.session._conn._conn.cursor() as cursor:
+                cursor.execute(f"drop table if exists {name}")
+            logging.debug(f"Cleanup Thread: Successfully dropped {common_log_text}")
+        except Exception:
+            logging.debug(
+                f"Cleanup Thread: Failed to drop {common_log_text}"
+            )  # pragma: no cover
+
+    def is_alive(self) -> bool:
+        return self.cleanup_thread is not None and self.cleanup_thread.is_alive()
+
+    def start(self) -> None:
+        if not self.is_alive():
+            self.cleanup_thread = threading.Thread(target=self.process_cleanup)
+            self.cleanup_thread.start()
+
+    def stop(self) -> None:
+        if self.is_alive():
+            self.queue.put(None)
+            self.cleanup_thread.join()
```

Comment:
so what happens when the clean up thread is stopped, and started, and during that time some table ref is dropped down. shouldn't we repick up clean up of those tables?

Reply:
with the current implementation, we will not be able to clean up those tables. I'm not sure whether it's really necessary because the above scenario is rare. But if we want to address this scenario, I'd like to 1) do not pop the table name from ref_count_map if it's zero, 2) make the cleanup thread run periodically to clean up all tables with zero count, instead of on-demand implementation with a queue, 3) only delete the table name in ref_count_map if it's dropped successfully.

Reply:
First of all, I am not sure if that scenario is rare or not, which seems an irrelevant question here. This is purely about the responsibility of the clean up thread, and the behavior of on thread start and stop. 

Based on the current implementation, it seems the current behavior is  
1) on stop, we basically will stop the clean up thread.
2) on start,  it will not start from where it left, so whatever table whose reference has reached to 0 during the time between stop and start will always be there.

If the clean up thread's responsibility is to clean up the table whose ref is 0 in the session, then we probably should resume the clean up. agree that we do not need to have a queue, but we may need some thread safe data structure, since you will have concurrent access to the map, since all concurrent write should be serialized by the main thread, this may not be bad.

if the cleanup thread's responsibility is to clean up the table whose ref becomes 0 after the thread starts, then we do not have to, which seems little bit incomplete to me.

I might leaning towards to the first behavior to get a compete table clean responsibility

Reply:
per our offline discussion, I changed the implementation to first behavior.